[DDP] Move gradient setting to None to python

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94258
* #94320

Some workloads are seeing not all grads set to None after backward
despite grad.reset() call in reducer, moving to the python callback alleviates
the issue.

Differential Revision: [D43071397](https://our.internmc.facebook.com/intern/diff/D43071397/)

rohan-varma committed:
[DDP] Move gradient setting to None to python

Some workloads are seeing not all grads set to None after backward
despite grad.reset() call in reducer, moving to the python callback alleviates
the issue.

Differential Revision: [D43071397](https://our.internmc.facebook.com/intern/diff/D43071397/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Move gradient setting to None to python"


Some workloads are seeing not all grads set to None after backward
despite grad.reset() call in reducer, moving to the python callback alleviates
the issue.

Differential Revision: [D43071397](https://our.internmc.facebook.com/intern/diff/D43071397/)

[ghstack-poisoned]

wanchaol commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -4609,7 +4609,7 @@ def _test_ddp_apply_optim_in_backward(
                 with torch.backends.cudnn.flags(
                     enabled=True, deterministic=True, benchmark=False
                 ):
-                    for i in range(100):
```
why we need to change the test iteration?

rohan-varma replied:
I initially added 100 to ensure determinism, but feel like 100 iterations increases TTS a bit too much. Can revert this back if needed.

wanchaol reviewed (approved):
stamp to unblock, please fix the lint before landing

awgu commented:
Do we know why calling `grad.reset()` in the reducer is not correctly setting gradients to `None`?

rohan-varma committed:
Update on "[DDP] Move gradient setting to None to python"


Some workloads are seeing not all grads set to None after backward
despite grad.reset() call in reducer, moving to the python callback alleviates
the issue.

Differential Revision: [D43071397](https://our.internmc.facebook.com/intern/diff/D43071397/)

[ghstack-poisoned]
